app/controllers/concerns/application_shared.rb

```diff
@@ -33,15 +33,17 @@ def set_taxonomy
       determined_taxonomy = nil
 
       if api_request? # API request
-        if user.admin? && !params.has_key?("#{taxonomy}_id") # admin always uses any context, they can limit the scope with explicit parameters such as organization_id(s)
-          determined_taxonomy = nil
-        elsif params.has_key?("#{taxonomy}_id") # admin and non-admin who specified explicit context
-          determined_taxonomy = available.where(:id => params["#{taxonomy}_id"]).first
+        if params.has_key?("#{taxonomy}_id") # admin and non-admin who specified context explicitly
+          if params["#{taxonomy}_id"].blank? # the key is present and explicitly set to nil which indicates "any" context
+            determined_taxonomy = nil
+          else
+            determined_taxonomy = available.where(:id => params["#{taxonomy}_id"]).first
 
-          # in case admin asked for taxonomy that does not exist or is not accessible, we reply with 404
-          if determined_taxonomy.nil?
-            not_found _("%{taxonomy} with id %{id} not found") % { :taxonomy => taxonomy.capitalize, :id => params["#{taxonomy}_id"] }
-            return false
+            # in case admin or non-admin asked for taxonomy that does not exist or is not accessible, we reply with 404
+            if determined_taxonomy.nil?
+              not_found _("%{taxonomy} with id %{id} not found") % { :taxonomy => taxonomy.capitalize, :id => params["#{taxonomy}_id"] }
```

Comment:
In our API I believe it is more common to issue 422 (unprocessable entity) in such case

Reply:
I'm not sure I understand, do you return 422 when object can't be found?

note that `not accessible` means privileges escalation and for unauthorized user we should return 404 to avoid information leak about not accessible org ids

Reply:
If I understand it correctly 404 would suite the situation where the API URI is not correct (e.g. when the non-existing org_id is part of the path) and the resource can't be found. If I get it right here it is about validity of params. The URI exists and the params use correct syntax though semantically wrong.

Excerpt from https://www.keycdn.com/support/422-unprocessable-entity/
```
The 422 (Unprocessable Entity) status code means the server 
understands the content type of the request entity 
(hence a 415(Unsupported Media Type) status code is inappropriate), 
and the syntax of the request entity is correct 
(thus a 400 (Bad Request) status code is inappropriate) 
but was unable to process the contained instructions.
```

Reply:
After the discussion I see the code is handling taxonomies in path as well as in params and there is no good way to distinguish them. 404 for both is okay then.